Building a native image does not work with vaadin-core-components from Vaadin 24.7

There have been changes in the components during the 24.7 cycle that makes them incompatible with Quarkus native images. In particular, https://github.com/vaadin/flow-components/pull/7011 adds references to optional Spring classes that cause the build to fail.

I did at some point also have an error with `com.vaadin.flow.component.shared.internal.OverlayAutoAddController` referenced from `com.vaadin.flow.component.confirmdialog.ConfirmDialog` but that might just have been a side effect from trying to work around the original issues.

Native compilation works if otherwise using Vaadin 24.6.6 but upgrading `vaadin-quarkus-extension` to version `24.7.0.beta2`.

Steps to reproduce:
1. Start from https://github.com/vaadin/base-starter-flow-quarkus
2. Change `vaadin.version` to `24.7.0.beta2` and add prerelease repositories
3. `mvn package -Pproduction -Dquarkus.package.type=native`

The build fails with this error:
```
Caused by: com.oracle.graal.pointsto.constraints.UnresolvedElementException: Discovered unresolved type during parsing: com.vaadin.flow.spring.data.VaadinSpringDataHelpers. This error is reported at image build time because class com.vaadin.flow.component.grid.Grid is registered for linking at image build time by command line and command line.
    at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.phases.SharedGraphBuilderPhase$SharedBytecodeParser.reportUnresolvedElement(SharedGraphBuilderPhase.java:604)
    at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.phases.SharedGraphBuilderPhase$SharedBytecodeParser.reportUnresolvedElement(SharedGraphBuilderPhase.java:598)
```

Tested with GraalVM CE 23.0.2+7.1

--- a/deployment/src/main/java/com/vaadin/quarkus/deployment/VaadinQuarkusNativeProcessor.java
+++ b/deployment/src/main/java/com/vaadin/quarkus/deployment/VaadinQuarkusNativeProcessor.java
@@ -15,7 +15,7 @@
  */
 package com.vaadin.quarkus.deployment;
 
-import jakarta.inject.Inject;
+import java.io.Serializable;
 import java.util.HashSet;
 import java.util.Objects;
 import java.util.Set;
@@ -42,6 +42,7 @@ import io.quarkus.gizmo.ClassCreator;
 import io.quarkus.gizmo.MethodCreator;
 import io.quarkus.undertow.deployment.ServletDeploymentManagerBuildItem;
 import io.quarkus.vertx.http.deployment.DefaultRouteBuildItem;
+import jakarta.inject.Inject;
 import org.atmosphere.cache.UUIDBroadcasterCache;
 import org.atmosphere.client.TrackMessageSizeInterceptor;
 import org.atmosphere.config.managed.ManagedServiceInterceptor;
@@ -193,6 +194,60 @@ public class VaadinQuarkusNativeProcessor {
                         .getConstructorCreator(new String[0]);
                 ctorCreator.setModifiers(Opcodes.ACC_PUBLIC);
                 ctorCreator.returnValue(null);
+            }
+        }
+    }
+
+    /*
+     * Some Vaadin Java components like Grid and Combobox define methods using
+     * VaadinSpringDataHelper utility from vaadin-spring module. In a Quarkus
+     * application that module is usually not present, so we provide a stub
+     * implementation that throws exception when methods are invoked.
+     */
+    @BuildStep(onlyIf = IsNativeBuild.class)
+    void generateVaadinSpringDataHelpers(
+            BuildProducer<GeneratedNativeImageClassBuildItem> producer) {
+        String vaadinSpringDataHelperClassName = "com.vaadin.flow.spring.data.VaadinSpringDataHelpers";
+
+        if (!QuarkusClassLoader
+                .isClassPresentAtRuntime(vaadinSpringDataHelperClassName)) {
+            ClassCreator.Builder builder = ClassCreator.interfaceBuilder()
+                    .className(vaadinSpringDataHelperClassName)
+                    .interfaces(Serializable.class.getName())
+                    .classOutput((className, bytes) -> producer.produce(
+                            new GeneratedNativeImageClassBuildItem(className,
+                                    bytes)));
+            try (ClassCreator classCreator = builder.build()) {
+
+                MethodCreator methodCreator = classCreator.getMethodCreator(
+                        "toSpringDataSort",
+                        "Lorg/springframework/data/domain/Sort;",
+                        "Lcom/vaadin/flow/data/provider/Query;");
+                methodCreator
+                        .setModifiers(Opcodes.ACC_PUBLIC | Opcodes.ACC_STATIC);
+                methodCreator.throwException(RuntimeException.class,
+                        "VaadinSpringDataHelpers.toSpringDataSort invoked but vaadin-spring is not available");
+                methodCreator.returnValue(null);
+
+                methodCreator = classCreator.getMethodCreator(
+                        "toSpringPageRequest",
+                        "Lorg/springframework/data/domain/PageRequest;",
+                        "Lcom/vaadin/flow/data/provider/Query;");
+                methodCreator
+                        .setModifiers(Opcodes.ACC_PUBLIC | Opcodes.ACC_STATIC);
+                methodCreator.throwException(RuntimeException.class,
+                        "VaadinSpringDataHelpers.toSpringPageRequest invoked but vaadin-spring is not available");
+                methodCreator.returnValue(null);
+
+                methodCreator = classCreator.getMethodCreator(
+                        "fromPagingRepository",
+                        "Lcom/vaadin/flow/data/provider/CallbackDataProvider$FetchCallback;",
+                        "Lcom/vaadin/flow/data/repository/PagingAndSortingRepository;");
+                methodCreator
+                        .setModifiers(Opcodes.ACC_PUBLIC | Opcodes.ACC_STATIC);
+                methodCreator.throwException(RuntimeException.class,
+                        "VaadinSpringDataHelpers.fromPagingRepository invoked but vaadin-spring is not available");
+                methodCreator.returnValue(null);
             }
         }
     }

